yt_dlp/cookies.py

```diff
@@ -0,0 +1,738 @@
+            else:
+                failed_message = ''
+            logger.info('extracted {} cookies from {}{}'.format(len(jar), browser_name, failed_message))
+            return jar
+        finally:
+            if cursor is not None:
+                cursor.connection.close()
+
+
+class ChromeCookieDecryptor:
+    """
+    Overview:
+
+        Linux:
+        - cookies are either v10 or v11
+            - v10: AES-CBC encrypted with a fixed key
+            - v11: AES-CBC encrypted with an OS protected key (keyring)
+            - v11 keys can be stored in various places depending on the activate desktop environment [2]
+
+        Mac:
+        - cookies are either v10 or not v10
+            - v10: AES-CBC encrypted with an OS protected key (keyring) and more key derivation iterations than linux
+            - not v10: 'old data' stored as plaintext
+
+        Windows:
+        - cookies are either v10 or not v10
+            - v10: AES-GCM encrypted with a key which is encrypted with DPAPI
+            - not v10: encrypted with DPAPI
+
+    Sources:
+    - [1] https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/os_crypt/
+    - [2] https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/os_crypt/key_storage_linux.cc
+        - KeyStorageLinux::CreateService
+    """
+
+    def decrypt(self, encrypted_value):
+        raise NotImplementedError
+
+
+def get_cookie_decryptor(browser_root, browser_keyring_name, logger):
+    if sys.platform in ('linux', 'linux2'):
+        return LinuxChromeCookieDecryptor(browser_keyring_name)
+    elif sys.platform == 'darwin':
+        return MacChromeCookieDecryptor(browser_keyring_name)
+    elif sys.platform == 'win32':
+        return WindowsChromeCookieDecryptor(browser_root, logger)
+    else:
+        raise NotImplementedError('Chrome cookie decryption is not supported '
+                                  'on this platform: {}'.format(sys.platform))
+
+
+class LinuxChromeCookieDecryptor(ChromeCookieDecryptor):
+    def __init__(self, browser_keyring_name):
+        self._v10_key = self.derive_key(b'peanuts')
+        if KEYRING_AVAILABLE:
+            self._v11_key = self.derive_key(_get_linux_keyring_password(browser_keyring_name))
+        else:
+            self._v11_key = None
+
+    @staticmethod
+    def derive_key(password):
+        # values from
+        # https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/os_crypt/os_crypt_linux.cc
+        return pbkdf2_sha1(password, salt=b'saltysalt', iterations=1, key_length=16)
+
+    def decrypt(self, encrypted_value):
+        version = encrypted_value[:3]
+        ciphertext = encrypted_value[3:]
+
+        if version == b'v10':
+            return _decrypt_aes_cbc(ciphertext, self._v10_key)
+
+        elif version == b'v11':
+            if self._v11_key is None:
+                warnings.warn('cannot decrypt cookie as the `keyring` module is not installed')
```

Comment:
the reason for using warnings here is so that only one warning is printed (because it will fail in the same way for every cookie). I could record some global state to only print once but that's not the nicest thing

Reply:
what about adding a parameter like `ytdl.report_warning(message, once=False)` then keep a record of the call sites of this method and only print the first call? this is what `warnings.warn` does internally:

https://github.com/python/cpython/blob/c89f0b2587eb0b16175a0bbb12d0b86314ff9320/Lib/warnings.py#L346-L349

Reply:
> decryption is guarenteed to fail always if it fails once

There are usually multiple encryption schemes present in a single database. I think the browser decides what to use based on how sensitive the cookie is. So even if decryption fails for some usable cookies may still be extracted

Reply:
I have added a param `only_once` for `report_warning`. Please replace all `warnings` with the Logger